controllers/tekton/pipeline/pipeline_controller.go

```diff
@@ -0,0 +1,195 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Pipeline object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *PipelineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	klog.Info(fmt.Sprintf("req: %v", req))
+
+	// get cr by its name
+	pipeline := &devopsv2alpha1.Pipeline{}
+	if err := r.Get(ctx, req.NamespacedName, pipeline); err != nil {
+		klog.Error(err, "unable to fetch pipeline crd resources")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// translate our pipeline CRD to tekton CRDs, e.g. Task, Pipeline, PipelineResource
+	if err := r.reconcileTektonCrd(ctx, req.Namespace, pipeline); err != nil {
+		klog.Error(err, "unable to reconcile tekton crd resources")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PipelineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&devopsv2alpha1.Pipeline{}).
+		Complete(r)
+}
+
+// reconcileTektonCrd translates our Pipeline CRD to Tekton CRDs
+func (r *PipelineReconciler) reconcileTektonCrd(ctx context.Context, namespace string, pipeline *devopsv2alpha1.Pipeline) error {
+	// print the pipeline name and the number of its tasks.
+	klog.Info(fmt.Sprintf("Pipeline name: %s\tTask num: %d", pipeline.Name, len(pipeline.Spec.Tasks)))
+
+	// translate tasks included in the pipeline to Tekton Tasks
+	klog.Info("Going to translate tasks included in the pipeline to Tekton Tasks.")
+	for _, task := range pipeline.Spec.Tasks {
+		if err := r.reconcileTektonTask(ctx, namespace, &task, pipeline.Name); err != nil {
+			klog.Error(err, "Failed to reconcile tekton task resources.")
+			return err
+		}
+	}
+
+	klog.Info("Going to translate Pipeline CRD to Tekton Pipeline CRD.")
+	if err := r.reconcileTektonPipeline(ctx, namespace, pipeline); err != nil {
+		klog.Error(err, "Failed to reconcile tekton pipeline resources.")
+		return err
+	}
+
+	return nil
+}
+
+// reconcileTektonTask translates tasks in our Pipeline CRD to Tekton Task CRD
+func (r *PipelineReconciler) reconcileTektonTask(ctx context.Context, namespace string, task *devopsv2alpha1.TaskSpec, pipelineName string) error {
+	// print the task name
+	klog.Info(fmt.Sprintf("Translating task %s to Tekton Task.", task.Name))
+
+	// check if the task already exists
+	// in order to differentiate tasks from pipelines, we add pipeline name before the task name
+	tTask := &tektonv1.Task{}
```

Comment:
I prefer to use `task`. And change `task *deevopsv2alpha1.TaskSpec` to `taskSepc`.

Reply:
Yes! I got confused when I was going to use `task`, becasuse `task` was already used in the function params. It's a good idea to revise `task *deevopsv2alpha1.TaskSpec` to `taskSpec`. Thanks!

Reply:
> I prefer to use `task`. And change `task *deevopsv2alpha1.TaskSpec` to `taskSepc`.

Yes! I got confused when I was going to use `task`, becasuse `task` was already used in the function params. It's a good idea to revise `task *deevopsv2alpha1.TaskSpec` to `taskSpec`. Thanks!